### PR TITLE
Atdd/failsafe feature

### LINKED ISSUE #763: ATDD-Staging: test for flow reroute when switch fails
Create a test that verifies successful flow reroute when transit switch turns down

### PR DESCRIPTION
**Pull request by rtretyak**
Added staging test for case when ISL fails.
Added request logger. Other refactoring

Closes #763

### CODE REVIEW

File: services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/service/flowcalculator/FlowCalculatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                List<PathInfoData> forwardPaths = topologyEngineService
+                        .getPaths(srcSwitch.getDpId(), dstSwitch.getDpId());
+                List<PathInfoData> reversePaths = topologyEngineService
+                        .getPaths(dstSwitch.getDpId(), srcSwitch.getDpId());
+                boolean hasAlternatePath = forwardPaths.size() > alternatePaths && reversePaths.size() > alternatePaths;
+                if (hasAlternatePath) {
+                    //try creating flow to see the actual path being used
+                    String flowId = format("%s-%s-%s", srcSwitch.getName(), dstSwitch.getName(),
+                            sdf.format(new Date()));
+                    FlowPayload flow = flowSet.new FlowBuilder(flowId, srcSwitch, dstSwitch)
```
Comment (sergii-iakovenko) [inline services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/service/flowcalculator/FlowCalculatorImpl.java]: Why not create a method in FlowSetBuilder? Something like "flowSet.buildFlowInUniqueVlan(flowId, srcSwitch, dstSwitch, bw)"

File: services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/service/flowcalculator/FlowCalculatorImpl.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+public class FlowCalculatorImpl implements FlowCalculator {
```
Comment (sergii-iakovenko) [inline services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/service/flowcalculator/FlowCalculatorImpl.java]: Will "FlowManager" fit better the purpose and role of the class?

File: services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/steps/FlowCrudSteps.java [hunk 3]
```diff
@@ -95,62 +97,42 @@
...
+    }
 
-                String flowId = format("%s-%s", srcSwitch.getName(), dstSwitch.getName());
-                builder.addFlow(flowId, srcSwitch, srcTraffGen.getSwitchPort(), dstSwitch, dstTraffGen.getSwitchPort());
-            });
-        });
+    @Given("Create (\\d+) flows? with A Switch used and at least (\\d+) alternate paths? between source and "
+            + "destination switch and (\\d+) bandwidth")
+    public void flowsWithAlternatePaths(int flowsAmount, int alternatePaths, int bw) {
+        topologyUnderTest.getFlowIsls().putAll(flowCalculator.createFlowsWithASwitch(flowsAmount, alternatePaths, bw));
```
Comment (sergii-iakovenko) [inline services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/steps/FlowCrudSteps.java]: 2 issues:
- It's better to split such invocation chains to improve readability. E.g. local variable.
- Bad practice to modify a collection obtained via getter.

File: services/src/atdd-staging/src/test/java/org/openkilda/atdd/staging/service/StubServiceFactory.java [hunk 4]
```diff
@@ -148,15 +154,33 @@ public FloodlightService getFloodlightStub() {
         return serviceMock;
     }
 
+    /**
+     * Get a stub for {@link ASwitchService}. The instance is tied to the factory state.
+     */
+    public ASwitchService getASwitchStub() {
+        ASwitchService serviceMock = mock(ASwitchService.class);
+        List<ASwitchFlow> lalaland = topologyDefinition.getIslsForActiveSwitches().stream()
```
Comment (sergii-iakovenko) [inline services/src/atdd-staging/src/test/java/org/openkilda/atdd/staging/service/StubServiceFactory.java]: Why "lalaland"?

File: services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/config/ServiceConfig.java [hunk 5]
```diff
@@ -76,30 +85,53 @@ public RetryPolicy retryPolicy() {
...
+    private RestTemplate buildRestTemplateWithAuth(String endpoint, String username, String password) {
+        RestTemplate restTemplate = buildLoggingRestTemplate(endpoint);
+        restTemplate.getInterceptors().add(new BasicAuthorizationInterceptor(username, password));
+        return restTemplate;
+    }
+
     private ResponseErrorHandler buildErrorHandler() {
         return new DefaultResponseErrorHandler() {
-            private final Logger LOGGER = LoggerFactory.getLogger(ResponseErrorHandler.class);
+            private final Logger logger = LoggerFactory.getLogger(ResponseErrorHandler.class);
```
Comment (sergii-iakovenko) [inline services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/config/ServiceConfig.java]: Make it static as well or use lombok @Slf4j.

File: services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/config/ServiceConfig.java [hunk 6]
```diff
@@ -76,30 +85,53 @@ public RetryPolicy retryPolicy() {
...
 
-    private RestTemplate buildRestTemplateWithAuth(String endpoint, String username, String password) {
-        final RestTemplate restTemplate = new RestTemplate();
+    @Bean(name = "aSwitchRestTemplate")
+    public RestTemplate aswitchtRestTemplate(@Value("${aswitch.endpoint}") String endpoint) {
+        return buildLoggingRestTemplate(endpoint);
+    }
+
+    @Bean
+    public FlowManager flowCalculator() {
```
[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/atdd-staging/src/main/java/org/openkilda/atdd/staging/config/ServiceConfig.java]: fixed. Forgot to rename along with FlowManager

**Pull request merged**